File: src/main/java/ru/citeck/ecos/notifications/domain/template/converter/TemplateConverter.java
package ru.citeck.ecos.notifications.domain.template.converter;

import lombok.RequiredArgsConstructor;
import org.jetbrains.annotations.NotNull;
import org.springframework.stereotype.Component;
import ru.citeck.ecos.commons.data.MLText;
import ru.citeck.ecos.commons.json.Json;
import ru.citeck.ecos.notifications.domain.template.dto.NotificationTemplateDto;
import ru.citeck.ecos.notifications.domain.template.dto.TemplateDataDto;
import ru.citeck.ecos.notifications.domain.template.entity.NotificationTemplate;
import ru.citeck.ecos.notifications.domain.template.entity.TemplateData;
import ru.citeck.ecos.notifications.domain.template.repository.NotificationTemplateRepository;

import java.util.HashMap;
import java.util.Map;

@Component
@RequiredArgsConstructor
public class TemplateConverter {

    private final NotificationTemplateRepository templateRepository;

    public NotificationTemplate dtoToEntity(@NotNull NotificationTemplateDto dto) {
        NotificationTemplate entity = templateRepository.findOneByExtId(dto.getId()).orElse(new NotificationTemplate());
        entity.setName(dto.getName());
        entity.setNotificationTitle(Json.getMapper().toString(dto.getNotificationTitle()));
        entity.setExtId(dto.getId());

        Map<String, TemplateData> updatedData = new HashMap<>();

        dto.getData().forEach((lang, dataDto) -> {
            TemplateData td = entity.getData().get(lang);
            if (td == null) {
                td = new TemplateData();
            }
            td.setLang(lang);
            td.setName(dataDto.getName());
            td.setData(dataDto.getData());
            td.setTemplate(entity);

            updatedData.put(lang, td);
        });

        entity.setData(updatedData);

        return entity;
    }

    public NotificationTemplateDto entityToDto(@NotNull NotificationTemplate entity) {
        NotificationTemplateDto dto = new NotificationTemplateDto();

        dto.setId(entity.getExtId());
        dto.setName(entity.getName());
        dto.setNotificationTitle(Json.getMapper().read(entity.getNotificationTitle(), MLText.class));

        Map<String, TemplateDataDto> data = new HashMap<>();
        entity.getData().forEach((lang, templateData) -> {
            TemplateDataDto dataDto = new TemplateDataDto();
            dataDto.setName(templateData.getName());
            dataDto.setData(templateData.getData());

            data.put(lang, dataDto);
        });

        dto.setData(data);

        return dto;
    }

}

File: src/main/java/ru/citeck/ecos/notifications/domain/template/converter/TemplateConverter.kt
package ru.citeck.ecos.notifications.domain.template.converter

import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Component
import ru.citeck.ecos.commons.data.MLText
import ru.citeck.ecos.commons.json.Json.mapper
import ru.citeck.ecos.notifications.domain.template.dto.MultiTemplateElementDto
import ru.citeck.ecos.notifications.domain.template.dto.NotificationTemplateWithMeta
import ru.citeck.ecos.notifications.domain.template.dto.TemplateDataDto
import ru.citeck.ecos.notifications.domain.template.repo.NotificationTemplateEntity
import ru.citeck.ecos.notifications.domain.template.repo.NotificationTemplateRepository
import ru.citeck.ecos.notifications.domain.template.repo.TemplateDataEntity
import java.util.*

private const val TAGS_SEPARATOR = ","

@Component
class TemplateConverter {

    @Autowired
    private lateinit var templateRepository: NotificationTemplateRepository

    fun dtoToEntity(dto: NotificationTemplateWithMeta): NotificationTemplateEntity {
        val entity = templateRepository.findOneByExtId(dto.id).orElse(NotificationTemplateEntity())
        entity.name = dto.name
        entity.notificationTitle = mapper.toString(dto.notificationTitle)
        entity.tags = dto.tags.joinToString(TAGS_SEPARATOR)
        entity.extId = dto.id
        entity.model = mapper.toString(dto.model)
        entity.multiTemplateConfig = mapper.toString(dto.multiTemplateConfig)

        val updatedData: MutableMap<String, TemplateDataEntity> = HashMap()

        dto.templateData.forEach { (lang: String, dataDto: TemplateDataDto) ->
            val td = entity.data[lang] ?: TemplateDataEntity()

            td.name = dataDto.name
            td.lang = lang
            td.data = dataDto.data
            td.template = entity

            updatedData[lang] = td
        }
        entity.data = updatedData
        return entity
    }

    fun entityToDto(entity: NotificationTemplateEntity): NotificationTemplateWithMeta {
        val dto = NotificationTemplateWithMeta(
            id = entity.extId!!,
            name = entity.name,
            notificationTitle = mapper.read(entity.notificationTitle, MLText::class.java),
            tags = if (entity.tags.isNullOrBlank()) {
                emptyList()
            } else {
                entity.tags!!.split(TAGS_SEPARATOR)
                    .map { it.trim() }
            },
            model = mapper.readMap(entity.model, String::class.java, String::class.java),
            multiTemplateConfig = mapper.readList(entity.multiTemplateConfig, MultiTemplateElementDto::class.java),
            creator = entity.createdBy,
            created = entity.createdDate,
            modifier = entity.lastModifiedBy,
            modified = entity.lastModifiedDate
        )

        val data: MutableMap<String, TemplateDataDto> = HashMap()

        entity.data.forEach { (lang: String, templateData: TemplateDataEntity) ->
            if (templateData.name != null && templateData.data != null) {
                val dataDto = TemplateDataDto(templateData.name!!, templateData.data!!)
                data[lang] = dataDto
            }
        }
        dto.templateData = data
        return dto
    }
}
